lib/navigation/preprocess_encoders.cpp

```diff
@@ -2,55 +2,99 @@
+  SensorChecks sensors    = checkReliable();
+  if (!encoder_data.has_value() || sensors == SensorChecks::kUnacceptable) {
+    return std::nullopt;
+  } else {
+    return encoder_data;
+  }
 }
 
-core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData encoder_data)
+std::optional<core::EncoderData> EncodersPreprocessor::detectOutliers(
+  const core::EncoderData encoder_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::EncoderData encoder_data_copy;
+  std::copy(encoder_data.begin(), encoder_data.end(), encoder_data_copy.begin());
+  Quartile quartiles;
+  core::Float interquartile_range;
+  core::Float upper_bound;
+  core::Float lower_bound;
+  // Case 1: All sensors are functioning properly
+  if (num_reliable_encoders_ == core::kNumEncoders) {
+    core::EncoderData reliable_data;
+    std::copy(encoder_data.begin(), encoder_data.end(), reliable_data.begin());
+    quartiles           = getQuartiles(reliable_data);
+    interquartile_range = quartiles.q3 - quartiles.q1;
+    upper_bound         = quartiles.median + 1.5 * interquartile_range;
+    lower_bound         = quartiles.median - 1.5 * interquartile_range;
+
+  }
+  // Case 2: One of the sensors has become faulty
+  else if (num_reliable_encoders_ == core::kNumEncoders - 1) {
+    std::array<uint32_t, core::kNumEncoders - 1> reliable_data;
+    std::size_t counter = 0;
+    for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+      if (encoders_reliable_.at(i)) {
+        reliable_data.at(counter) = encoder_data.at(i);
+        ++counter;
+      }
+    }
+    quartiles           = getQuartiles(reliable_data);
+    interquartile_range = quartiles.q3 - quartiles.q1;
+    upper_bound         = quartiles.median + 1.2 * interquartile_range;
+    lower_bound         = quartiles.median - 1.2 * interquartile_range;
+  }
+  // Case 3: More than one sensors have become unreliable, stopping everything and entering into
+  // fail State.
+  else {
+    logger_.log(core::LogLevel::kFatal,
+                "Number of unreliable encoder sensors have exceeded the threshold");
+    return std::nullopt;
+  }
+
+  for (std::size_t i = 0; i < encoder_data_copy.size(); ++i) {
+    // replacing the ouliers or data from faulty sensors with the median of the dataset.
+    if (encoder_data_copy.at(i) > upper_bound || encoder_data_copy.at(i) < lower_bound
+        || encoders_reliable_.at(i) == false) {
+      encoder_data_copy.at(i) = quartiles.median;
+
+      if (encoders_reliable_.at(i)) { ++encoder_outliers_.at(i); }
+    } else {
+      encoder_outliers_.at(i) = 0;
+    }
```

Comment:
this could probably be structured better, seems weird to have one outer condition being `encoders_reliable_.at(i) == false` then also checking within that condition if `encoders_reliable_.at(i) == true`

Reply:
This extra statement is necessary because otherwise if we get one failure, the system will keep counting that failure on every following iteration, leading to an unnecessary critical failure...